### PR TITLE
[WIP] P2P rudimentary scaling

### PR DESCRIPTION
**Pull request by bfbachmann**
# What changed:
- Added `message` package
- Added `subnet` package
- Added `errors` package
- Refactor and add some functionality to `Peer` package

### CODE REVIEW

File: message/message.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+// Message types
+// NOTE: because of the way iota works, changing the order in which the
+// following constants appear will change their values, which may affect the
+// ability of your peer to communicate with others.
+const (
+	// Send the multiaddress of a peer to another peer
+	PeerInfo = iota
+	// Request addressess of peers in the remote peer's subnet
+	RequestPeerInfo = iota
```
Comment (jordanschalm) [inline message/message.go]: What is the difference between these two? When would we ever need only one peer's info

File: message/message.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	"strings"
+)
+
+// Message types
+// NOTE: because of the way iota works, changing the order in which the
+// following constants appear will change their values, which may affect the
+// ability of your peer to communicate with others.
+const (
+	// Send the multiaddress of a peer to another peer
+	PeerInfo = iota
```
Comment (jordanschalm) [inline message/message.go]: You only need to write iota the first time
Like:
```
const (
  a = iota
  b
  c
)
// a = 0, b = 1, c = 2
```

File: message/message.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	// Request addressess of peers in the remote peer's subnet
+	RequestPeerInfo = iota
+	// Send information about a block that was just hashed
+	NewBlock = iota
+	// Request chunk of the blockchain from peer
+	RequestChunk = iota
+	// Advertise that we have a chunk of the blockchain
+	AdvertiseChunk = iota
```
Comment (jordanschalm) [inline message/message.go]: I think we should assume that every node has the whole blockchain. That should be the first thing they do.

File: message/message.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+// Bytes returns JSON representation of message as a byte array, or error if
+// message cannot be marshalled.
+func (m *Message) Bytes() ([]byte, error) {
+	return json.Marshal(m)
+}
+
+// FromString parses a message in the form of a string and returns a pointer
+// to a new Message struct made from the contents of the string. Returns error
+// if string is malformed.
+func FromString(s string) (*Message, error) {
```
Comment (jordanschalm) [inline message/message.go]: Should this be `FromBytes`/`Marshal` instead? Why is it a string?

File: peer/peer.go [hunk 5]
```diff
@@ -89,38 +93,99 @@ func New(ip string, port int) (*Peer, error) {
...
+
+// Broadcast sends message to all peers this peer is currently connected to
+func (p *Peer) Broadcast(m msg.Message) error {
+	return errors.New("Function not implemented")
 }
 
 // ExtractPeerInfo extracts the peer ID and multiaddress from the
 // given multiaddress.
 // Returns peer ID (esentially 46 character hash created by the peer)
 // and the peer's multiaddress in the form /ip4/<peer IP>/tcp/<CumulusPort>.
```
Comment (bfbachmann) [inline peer/peer.go:299]: I'm not really sure that this function is necessary. I feel like it's doing much more work than it has to.

File: message/message.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+		return nil, errors.New("Invalid message type")
+	}
+
+	m := &Message{msgType: t, content: c}
+	return m, nil
+}
+
+// Bytes returns JSON representation of message as a byte array, or error if
+// message cannot be marshalled.
+func (m *Message) Bytes() ([]byte, error) {
```
Comment (jordanschalm) [inline message/message.go]: `Marshal` for consistency

Comment (jordanschalm) [reply message/message.go]: Use `gob` for en/decoding

File: peer/peer.go [hunk 7]
```diff
@@ -89,38 +93,99 @@ func New(ip string, port int) (*Peer, error) {
...
-// This should be passed as the second argument to SetStreamHandler().
-// We may want to implement another type of StreamHandler in the future.
+// opens a new stream with this peer.
+// This should be passed as the second argument to SetStreamHandler() after this
+// peer is initialized.
 func (p *Peer) Receive(s net.Stream) {
-	log.Debug("Setting basic stream handler.")
-	p.doCumulus(s)
-}
+	defer s.Close()
```
Comment (jordanschalm) [inline peer/peer.go]: Do you want to close the stream every time you receive? Is it a new stream each time?

File: peer/peer.go [hunk 8]
```diff
@@ -89,38 +93,99 @@ func New(ip string, port int) (*Peer, error) {
...
+
+	// Get remote peer's full multiaddress
+	remoteMA, err := makeMultiaddr(
+		s.Conn().RemoteMultiaddr(), s.Conn().RemotePeer())
+	if err != nil {
+		log.Fatal("Failed to obtain valid remote peer multiaddress")
+	}
+
+	// Add the remote peer to this peer's subnet
+	err = p.subnet.AddPeer(remoteMA, s)
```
Comment (jordanschalm) [inline peer/peer.go]: What if the peer is already in our subnet? I feel like `Receive` should function like:
* unmarshal message
* switch on message type and call a subroutine

File: peer/peer.go [hunk 9]
```diff
@@ -89,38 +93,99 @@ func New(ip string, port int) (*Peer, error) {
...
+	}
+	defer p.subnet.RemovePeer(remoteMA)
 
-// Communicate with peers.
-// TODO: Update this to do something useful. For now it just reads from the
-// stream and writes back what it read.
-func (p *Peer) doCumulus(s net.Stream) {
 	buf := bufio.NewReader(s)
-	str, err := buf.ReadString('\n')
+	strMsg, err := buf.ReadString('\n') // TODO: set timeout here
```
Comment (jordanschalm) [inline peer/peer.go]: Just get the bytes from the buffer here and add `unmarshal from byte slice` in the message object

File: peer/peer.go [hunk 10]
```diff
@@ -89,38 +93,99 @@ func New(ip string, port int) (*Peer, error) {
...
 	if err != nil {
 		log.Error(err)
```
Comment (jordanschalm) [inline peer/peer.go]: You can do `log.WithError(err).Error("Error doing whatever")

File: peer/peer.go [hunk 11]
```diff
@@ -14,23 +14,25 @@ import (
...
 	CumulusProtocol = "/cumulus/0.0.1"
-
 	// DefaultIP is the IP address new hosts will use if none if provided
 	DefaultIP = "127.0.0.1"
 )
 
 // Peer is a cumulus Peer composed of a host
 type Peer struct {
 	host.Host
+	subnet sn.Subnet
```
Comment (chadlagore) [inline peer/peer.go]: I _thiiiiink_ we're using capital letters, see https://github.com/ubclaunchpad/cumulus/pull/23/files `Subnet`

File: peer/peer.go [hunk 12]
```diff
@@ -150,25 +215,48 @@ func ExtractPeerInfo(peerma string) (lpeer.ID, ma.Multiaddr, error) {
...
-// Connect adds the given multiaddress to p's Peerstore and opens a stream
-// with the peer at that multiaddress if the multiaddress is valid, otherwise
-// returns error.
-func (p *Peer) Connect(peerma string) (net.Stream, error) {
-	peerid, targetAddr, err := ExtractPeerInfo(peerma)
-	if err != nil {
-		return nil, err
+// advertisePeers writes messages into the given stream advertising the
+// multiaddress of each peer in this peer's subnet.
+func (p *Peer) advertisePeers(s net.Stream) {
```
Comment (jordanschalm) [inline peer/peer.go]: Why do we need to advertise peers? Why not just give our list of peers during connection, or accept requests for our list of peers. Seems like an an unnecessary complication

File: peer/peer.go [hunk 13]
```diff
@@ -150,25 +215,48 @@ func ExtractPeerInfo(peerma string) (lpeer.ID, ma.Multiaddr, error) {
...
+	mAddrs := p.subnet.Multiaddrs()
+	log.Debug("Peers on this subnet: ")
+	for mAddr := range mAddrs {
```
Comment (jordanschalm) [inline peer/peer.go]: This seems like it should be refactored into `Broadcast`

File: peer/peer.go [hunk 14]
```diff
@@ -150,25 +215,48 @@ func ExtractPeerInfo(peerma string) (lpeer.ID, ma.Multiaddr, error) {
...
+		mAddrString := string(mAddr)
+		log.Debug("\t", mAddrString)
+		message, err := msg.New([]byte(mAddrString), msg.PeerInfo)
+		if err != nil {
+			log.Error("Failed to create message")
```
Comment (jordanschalm) [inline peer/peer.go]: Same as above. Use `log.WithError(....)` Also two instances below

File: peer/peer.go [hunk 15]
```diff
@@ -150,25 +215,48 @@ func ExtractPeerInfo(peerma string) (lpeer.ID, ma.Multiaddr, error) {
...
-	log.Debug("Peer ID:", peerid.Pretty())
-	log.Debug("Peer Address:", targetAddr)
-
-	// Open a stream with the peer
-	stream, err := p.NewStream(context.Background(), peerid,
-		CumulusProtocol)
+// makeMultiaddr creates a Multiaddress from the given Multiaddress (of the form
+// /ip4/<ip address>/tcp/<TCP port>) and the peer id (a hash) and turn them
+// into one Multiaddress. Will return error if Multiaddress is invalid.
+func makeMultiaddr(iAddr ma.Multiaddr, pid lpeer.ID) (ma.Multiaddr, error) {
```
Comment (jordanschalm) [inline peer/peer.go]: Use `NewMultiAddr` for consistency

File: peer/peer.go [hunk 16]
```diff
@@ -150,25 +215,48 @@ func ExtractPeerInfo(peerma string) (lpeer.ID, ma.Multiaddr, error) {
...
+	strMA := fmt.Sprintf("%s/ipfs/%s", strAddr, strID)
+	mAddr, err := ma.NewMultiaddr(strMA)
+	return mAddr, err
+}
 
-	return stream, err
+func (p *Peer) handleMessage(m msg.Message, s net.Stream) {
+	switch m.Type() {
+	case msg.RequestPeerInfo:
+		p.advertisePeers(s)
```
Comment (jordanschalm) [inline peer/peer.go]: Why advertise to everyone when only one peer requests?

File: subnet/subnet.go [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+	}
+	return nil
+}
+
+// RemovePeer removes the mapping with the key mAddr from the subnet if it
+// exists.
+func (sn *Subnet) RemovePeer(mAddr ma.Multiaddr) {
+	log.Debugf("Removing peer %s from subnet", mAddr.String())
+	delete(sn.peers, mAddr)
+	sn.numPeers--
```
Comment (jordanschalm) [inline subnet/subnet.go]: Don't separately keep track of `numPeers`, just use `len(sn.peers)`. It's O(1)

File: subnet/subnet.go [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+// isFull returns true if the number of peers in the sunbet is at or over the
+// limit set for that subnet, otherwise returns false.
+func (sn *Subnet) isFull() bool {
```
Comment (jordanschalm) [inline subnet/subnet.go]: `Full` not `isFull`

**Review by jordanschalm — CHANGES_REQUESTED** [review]
Highest Quality 👌

**Review by chadlagore — APPROVED** [review]

File: message/message.go (hunk 4 above)
Comment (bfbachmann) [reply message/message.go]: Yea I think this should be `Unmarshal` and should take a `[]byte` as an argument.

File: peer/peer.go (hunk 7 above)
Comment (bfbachmann) [reply peer/peer.go]: This function will only terminate when we are fully finished communicating with the peer on the other end of the stream. That functionality has yet to be implemented.

File: peer/peer.go (hunk 8 above)
Comment (bfbachmann) [reply peer/peer.go]: `AddPeer` handles the case where the peer is already in our subnet. I think your description of what `Receive` should do hits the nail on the head.

File: peer/peer.go (hunk 12 above)
Comment (bfbachmann) [reply peer/peer.go]: Yea I guess this should be called `sendPeerInfo`. It should only get called when a peer requests peer info.

File: subnet/subnet.go (hunk 17 above)
Comment (bfbachmann) [reply subnet/subnet.go]: Wow yea that is just disgusting.

File: peer/peer.go (hunk 11 above)
Comment (bfbachmann) [reply peer/peer.go]: If it's capitalized its public, so I decided against it.

**Pull request closed without merging**